File: LeetCodeSolutions/LeetCode_0315.py
class FenwickTree:
    def __init__(self, n):
        self.size = n
        self.tree = [0] * (n + 1)

    def __lowbit(self, index):
        return index & (- index)

    def update(self, index, delta):
        while index < self.size + 1:
            self.tree[index] += delta
            index += self.__lowbit(index)

    def query(self, index):
        res = 0
        while index > 0:
            res += self.tree[index]
            index -= self.__lowbit(index)
        return res


class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:
        n = len(nums)
        if n < 1:
            return []
        if n == 1:
            return [0]
        s = list(set(nums))
        total_rank = len(s)
        rank_map = {}
        import heapq
        heapq.heapify(s)
        for _ in range(1, total_rank + 1):
            rank_map[heapq.heappop(s)] = _
        res = []
        ft = FenwickTree(total_rank)
        for _ in range(n - 1, -1, -1):
            index = rank_map[nums[_]]
            ft.update(index, 1)
            res = [ft.query(index - 1)] + res
        return res

File: LeetCodeSolutions/LeetCode_0315.cpp
class BIT{
    public:
    vector<int> bit;
    BIT(int n){
        bit.resize(n, 0);
    }
    inline int lowbit(int x){return x & -x;}
    int query(int x){
        int ret = 0;
        x -= 1;
        while(x  >= 1){
            ret += bit[x];
            x -= lowbit(x);
        }
        return ret;
    }

    void update(int x){
        while (x < bit.size()){
            ++bit[x];
            x += lowbit(x);
        }
    }

};

class Solution {
public:
    vector<int> countSmaller(vector<int>& nums) {
        vector<int> tmp(nums.begin(), nums.end());
        sort(tmp.begin(), tmp.end());
        unordered_map<int, int> n2id;
        int cnt = 1;
        for(int i = 0; i < tmp.size(); ++i){
            if (i > 0 && tmp[i] == tmp[i - 1]) continue;
            n2id[tmp[i]] = cnt++;
        }
        BIT bit(cnt);
        vector<int> ret;
        for(int i = nums.size() - 1; i >=0; --i){
            int uid = n2id[nums[i]];
            ret.push_back(bit.query(uid));
            bit.update(uid);
        }
        return {ret.rbegin(), ret.rend()};

    }
};